tests/FetchAuthTokenTest.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+namespace Google\Auth\Tests;
+
+use Google\Auth\Credentials\AppIdentityCredentials;
+use Google\Auth\Credentials\GCECredentials;
+use Google\Auth\Credentials\ServiceAccountCredentials;
+use Google\Auth\Credentials\ServiceAccountJwtAccessCredentials;
+use Google\Auth\Credentials\UserRefreshCredentials;
+use Google\Auth\CredentialsLoader;
+use Google\Auth\FetchAuthTokenInterface;
+use Google\Auth\OAuth2;
+
+class FetchAuthTokenTest extends BaseTest
+{
+  /** @dataProvider provideAuthTokenFetcher */
+  public function testGetLastReceivedToken(FetchAuthTokenInterface $fetcher)
+  {
+    $accessToken = $fetcher->getLastReceivedToken();
+
+    $this->assertNotNull($accessToken);
+    $this->assertArrayHasKey('access_token', $accessToken);
+    $this->assertArrayHasKey('expires', $accessToken);
+
+    $this->assertEquals('xyz', $accessToken['access_token']);
+    $this->assertEquals(strtotime('2001'), $accessToken['expires']);
+  }
+
+  public function provideAuthTokenFetcher()
+  {
+    $scopes = ['https://www.googleapis.com/auth/drive.readonly'];
+    $jsonPath = __DIR__ . '/fixtures/.config/' . CredentialsLoader::WELL_KNOWN_PATH;
+    $jsonPath2 = __DIR__ . '/fixtures2/.config/' . CredentialsLoader::WELL_KNOWN_PATH;
+
+    return [
+      [ $this->getAppIdentityCredentials() ],
+      [ $this->getGCECredentials() ],
+      [ $this->getServiceAccountCredentials($scopes, $jsonPath) ],
+      [ $this->getServiceAccountJwtAccessCredentials($jsonPath) ],
+      [ $this->getUserRefreshCredentials($scopes, $jsonPath2) ],
+      [ $this->getOAuth2() ],
+    ];
+  }
+
+  private function getAppIdentityCredentials()
+  {
+    $class = new \ReflectionClass('Google\Auth\Credentials\AppIdentityCredentials');
+    $property = $class->getProperty('lastReceivedToken');
+    $property->setAccessible(true);
+
+    $credentials = new AppIdentityCredentials();
+    $property->setValue($credentials, [
+      'access_token' => 'xyz',
+      'expiration_time' => strtotime('2001'),
+    ]);
+
+    return $credentials;
+  }
+
+  private function getGCECredentials()
+  {
+    $class = new \ReflectionClass('Google\Auth\Credentials\GCECredentials');
+    $property = $class->getProperty('lastReceivedToken');
+    $property->setAccessible(true);
+
+    $credentials = new GCECredentials();
+    $property->setValue($credentials, [
+      'access_token' => 'xyz',
+      'expiration_time' => strtotime('2001'),
+    ]);
+
+    return $credentials;
+  }
+
+  private function getServiceAccountCredentials($scopes, $jsonPath)
+  {
+    $class = new \ReflectionClass('Google\Auth\Credentials\ServiceAccountCredentials');
```

Comment:
Not sure what the character limit you guys are using for your project is, so not sure if this line length is okay or needs to be fixed.